Sources/TensorFlow/Operators/Basic.swift

```diff
@@ -390,44 +390,92 @@ public extension Tensor {
-            to: Tensor<Index>(dValue),
-            stride: Tensor<Index>(ones: [])
-        ) * accumulated
-        let dShape = Tensor<Int32>(concatenating: [
-            dValue.rankLifted(),
-            Tensor<Int32>([Int32](repeating: 1, count: indices.rank - 1))])
-        batchIndices += dIndices.reshaped(toShape: dShape)
+    func batchGathering<Index: TensorFlowIndex>(
+        atIndices indices: Tensor<Index>,
+        alongAxis axis: Int = 1,
+        batchDimCount batchDims: Int = 1
+    ) -> Tensor {
+        // TODO: precondition(batchDims >= 0 && batchDims < indices.rank,
+        //                    "'numBatchDims' must be non-negative and less than 'indices.rank'.")
+        // TODO: precondition(batchDims < rank, 
+        //                    "'numBatchDims' must be less than the tensor's rank.")
+
+        // Handle the axis argument by transposing the axis dimension so that it is the first
+        // non-batch dimension, recursively calling `batchGathering` with `axis = 0`, and then
+        // transposing the result to put the pre-axis dimensions before the indices dimensions.
+        if axis != batchDims {
+            // Adjust axis to be positive.
+            let posAxis = axis < 0 ? axis + rank : axis
+
+            // TODO: precondition(posAxis >= 0 && posAxis < rank, "'axis' is out of range.")
+            // TODO: precondition(batchDims <= posAxis,
+            //                    "'batchDims' must be less than or equal to 'axis'.")
+
+            // Move self[axis] up to self[batchDims].
+            let permutation = Tensor<Int32>(concatenating: [
+                Tensor<Int32>(rangeFrom: 0, to: Int32(batchDims), stride: 1),
+                Tensor<Int32>(Int32(axis)).rankLifted(),
+                Tensor<Int32>(rangeFrom: Int32(batchDims), to: Int32(posAxis), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(axis) + 1, to: Int32(rank), stride: 1)])
+            let tensor = transposed(withPermutations: permutation)
+            let result = tensor.batchGathering(
+                atIndices: indices,
+                alongAxis: batchDims,
+                batchDimCount: batchDims)
+
+            // Move the result dimensions corresponding to self[batchDims ..< axis] to just before
+            // the dimensions corresponding to indices[batchDims ...].
+            let start = indices.rank + posAxis - batchDims
+            let resultPermutation = Tensor<Int32>(concatenating: [
+                Tensor<Int32>(rangeFrom: 0, to: Int32(batchDims), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(indices.rank), to: Int32(start), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(batchDims), to: Int32(indices.rank), stride: 1),
+                Tensor<Int32>(rangeFrom: Int32(start), to: Int32(result.rank), stride: 1)])
+            return result.transposed(withPermutations: resultPermutation)
+        }
+
+        let batchIndices = withoutDerivative(at: { () -> Tensor<Index> in
```

Comment:
```suggestion
        let batchIndices = withoutDerivative {
```

Reply:
`withoutDerivative(at:)` does not take a closure as argument and so I cannot use the trailing closure syntax.

Reply:
It's generic over `T`, so `T` will be inferred to be any trailing closure. Did that not work for you?

Reply:
I get an error. I believe that's because I don't want to pass the closure, but rather the value it returns and I wrap the whole thing in `withoutDerivative(at:)`. Otherwise, if I make the suggested change I would have to treat `batchIndices` as a function.